metaverse_modules/heal-effect/index.js

```diff
@@ -0,0 +1,426 @@
+                          swAttribute.setX(i,0.95);
+                          scalesAttribute.setX(i, 4.9);
+                          if(!circlePlay)
+                              circlePlay = true;
+                      }
+                      break;
+                  }
+                  case 2: {
+                      if(scalesAttribute.getX(i)<4){
+                          if(swAttribute.getX(i)>=1)
+                              scalesAttribute.setX(i, scalesAttribute.getX(i)+0.9);
+                          else{
+                              if(scalesAttribute.getX(i)>0)
+                                  scalesAttribute.setX(i, scalesAttribute.getX(i)-0.8);
+                              else
+                                  scalesAttribute.setX(i, 0);
+                          }
+                      }
+                      else{
+                          swAttribute.setX(i,0.95);
+                          scalesAttribute.setX(i, 3.9);
+                      }
+                      break;
+                  }
+              }
+          }
+        
+          idAttribute.needsUpdate = true;
+          positionsAttribute.needsUpdate = true;
+          swAttribute.needsUpdate = true;
+          scalesAttribute.needsUpdate = true;
+          flashMesh.material.uniforms.cameraBillboardQuaternion.value.copy(camera.quaternion);
+          flashMesh.material.uniforms.avatarPos.x=localPlayer.position.x;
+          flashMesh.material.uniforms.avatarPos.y=localPlayer.position.y;
+          flashMesh.material.uniforms.avatarPos.z=localPlayer.position.z;
+
+          //#################################### handle pixel #######################################
+          for(let i = 0; i < pixelParticleCount; i++){
+              if(pixelOpacityAttribute.getX(i)>0){
+                  pixelScaleAttribute.setX(i,pixelScaleAttribute.getX(i)-0.01);
+                  pixelOpacityAttribute.setX(i,pixelOpacityAttribute.getX(i)-(0.01+Math.random()*0.03));
+                  pixelPositionAttribute.setY(i,pixelPositionAttribute.getY(i)+0.02);
+              }
+                  
+              else
+                  pixelOpacityAttribute.setX(i,0);
+          }
+          pixelOpacityAttribute.needsUpdate = true;
+          pixelPositionAttribute.needsUpdate = true;
+          pixelScaleAttribute.needsUpdate = true;
+          group.position.copy(localPlayer.position);
+
+          if(materialStartTime>0){
+            for(let i=0;i<healMaterial.length;i++){
+                if(healMaterial[i].g>0){
+                  // healMaterial[i].r += 0.02;
+                  healMaterial[i].g -= 0.02;
+                  // healMaterial[i].b += 0.02;
+                }
+                else{
+                  healMaterial[i].r = 0;
+                  healMaterial[i].g = 0;
+                  healMaterial[i].b = 0;
+                  app.remove(group);
+                  app.remove(flashMesh);
+                  particleAlreadyInScene=false;
```

Comment:
This isn't the best logic.

This is in the middle of the for loop for several materials, but it is setting app state (multiple times), and also removing apps multiple times.

If you flip the code and put the `for` inside of `if` it will be better.

Reply:
remove group and flashMesh once: https://github.com/webaverse/app/pull/2873/commits/a9a81a8ff1878f514d9df2da555ef826b24df7eb